pypit/arproc.py

```diff
@@ -857,46 +882,96 @@ def reduce_echelle(slf, sciframe, scidx, fitsdict, det, standard=False):
-    orders = 1.0 + np.arange(nord)
-    msgs.info("Performing a PCA on the object trace")
-    ofit = settings.argflag['trace']['object']['pca']['params']
-    lnpc = len(ofit) - 1
-    msgs.work("May need to do a check here to make sure ofit is reasonable")
-    xcen = trcxfit[:, np.newaxis].repeat(nord, axis=1)
-    trccen = arutils.func_val(trccoeff, trcxfit, settings.argflag['trace']['object']['function'],
-                              minv=0.0, maxv=nspec-1.0)
-    fitted, outpar = arpca.basis(xcen, trccen, trccoeff, lnpc, ofit, skipx0=False,
-                                 function=settings.argflag['trace']['object']['function'])
+    refine = 0.0
+    if settings.argflag['trace']['object']['method'] == "pca":
+        # Identify the orders to be extrapolated during reconstruction
+        orders = 1.0 + np.arange(nord)
+        msgs.info("Performing a PCA on the object trace")
+        ofit = settings.argflag['trace']['object']['params']
+        lnpc = len(ofit) - 1
+        msgs.work("May need to do a check here to make sure ofit is reasonable")
+        xcen = trcxfit[:, np.newaxis].repeat(nord, axis=1)
+        trccen = arutils.func_val(trccoeff, trcxfit, settings.argflag['trace']['object']['function'],
+                                  minv=0.0, maxv=nspec-1.0).T
+        fitted, outpar = arpca.basis(xcen, trccen, trccoeff, lnpc, ofit, skipx0=False,
+                                     function=settings.argflag['trace']['object']['function'])
+        if not msgs._debug['no_qa']:
+            arpca.pc_plot(slf, outpar, ofit, pcadesc="PCA of object trace")
+        # Extrapolate the remaining orders requested
+        trccen, outpar = arpca.extrapolate(outpar, orders, function=settings.argflag['trace']['object']['function'])
+        #refine = trccen-trccen[nspec//2, :].reshape((1, nord))
+    else:
+        msgs.error("Not ready for object trace method:" + msgs.newline() +
+                   settings.argflag['trace']['object']['method'])
+    # Construct the left and right traces of the object profile
+    # The following code ensures that the fraction of the slit
+    # containing the object remains constant along the spectral
+    # direction
+    trcmean = np.mean(trccen, axis=0)
+    trobjl = (trcmean - (1+bgnl)/slf._pixwid[det - 1].astype(np.float)).reshape((1, nord)).repeat(nspec, axis=0)
+    trobjl = trccen - trobjl
+    trobjr = (-trcmean + (slf._pixwid[det - 1]-bgnr-1)/slf._pixwid[det - 1].astype(np.float)).reshape((1, nord)).repeat(nspec, axis=0)
+    trobjr = trccen + trobjr
+    # Convert trccen to the actual trace locations
+    trccen *= (slf._rordloc[det - 1] - slf._lordloc[det - 1])
+    trccen += slf._lordloc[det - 1]
+    trobjl *= (slf._rordloc[det - 1] - slf._lordloc[det - 1])
+    trobjl += slf._lordloc[det - 1]
+    trobjr *= (slf._rordloc[det - 1] - slf._lordloc[det - 1])
+    trobjr += slf._lordloc[det - 1]
+
+    # Generate an image of pixel weights for each object
```

Comment:
Maybe specify what the weight factor ranges from.
Are they mainly 0's and 1's ?